amm-executor/src/Spectrum/Executor/Backlog/Service.hs

```diff
@@ -1,60 +1,158 @@
-mkBacklog MakeLogging{..} BacklogConfig{..} = do
-  logging     <- forComponent "Pools"
+  -> BacklogServiceConfig
+  -> BacklogStore m
+  -> f (BacklogService m)
+mkBacklogService MakeLogging{..} config store@BacklogStore{..} = do
+  logging     <- forComponent "BacklogService"
   -- those queues should be shared with Backlog.Proceess (to make live updates). So maybe worth extracting them into separate module. e.g. BacklogStore
   pendingPQ   <- newIORef PQ.empty  -- ordered by weight; new orders
   suspendedPQ <- newIORef PQ.empty  -- ordered by weight; failed orders, waiting for retry (retries are performed with some constant probability, e.g. 5%) 
   toRevisitQ  <- newIORef Seq.empty -- regular queue; successully submitted orders. Left orders should be re-executed in X minutes. Normally successfully confirmed orders are eliminated from this queue.
+  _           <- liftK $ recover store config pendingPQ
+  pure $ attachLogging logging BacklogService
+    { put = \(PendingOrder order timestamp) -> do
+        put $ BacklogOrder timestamp order
+        modifyIORef pendingPQ (PQ.insert $ mkWeightedOrderWithTimestamp order timestamp)
+    , suspend = \(SuspendedOrder order timestamp) -> do
+        modifyIORef suspendedPQ (PQ.insert $ mkWeightedOrderWithTimestamp order timestamp)
+        exists $ BacklogOrder timestamp order
+    , checkLater = \(InProgressOrder order timestamp) -> do
+        modifyIORef toRevisitQ (mkWeightedOrderWithTimestamp order timestamp Seq.<|)
+        exists $ BacklogOrder timestamp order
+    , tryAcquire = do
+        _ <- refreshQueues config store pendingPQ suspendedPQ toRevisitQ
+        maxWeightedOrderM <- getMaxWeightedOrder' pendingPQ suspendedPQ
+        case maxWeightedOrderM of
+          Just (WeightedOrderWithTimestamp oId _ _) -> get oId <&> fmap backlogOrder
+          Nothing -> pure Nothing
+    , drop = \oId -> do
+        _ <- modifyIORef pendingPQ   (PQ.filter (\(WeightedOrderWithTimestamp orderId _ _) -> orderId /= oId))
+        _ <- modifyIORef suspendedPQ (PQ.filter (\(WeightedOrderWithTimestamp orderId _ _) -> orderId /= oId))
+        _ <- modifyIORef toRevisitQ  (Seq.filter (\(WeightedOrderWithTimestamp orderId _ _) -> orderId /= oId))
+        drop oId
+    }
 
-  (_, db) <- Rocks.openBracket storePath
-              Rocks.defaultOptions
-                { Rocks.createIfMissing = createIfMissing
-                }
+refreshQueues
+  :: (MonadIO m)
+  => BacklogServiceConfig
+  -> BacklogStore m
+  -> IORef (PQ.MaxQueue WeightedOrderWithTimestamp)
+  -> IORef (PQ.MaxQueue WeightedOrderWithTimestamp)
+  -> IORef (Seq.Seq WeightedOrderWithTimestamp)
+  -> m ()
+refreshQueues BacklogServiceConfig{..} BacklogStore{..} pendingQueueRef suspendedQueueRef toRevisitSeqRef = do
+  currentTime <- getCurrentTime
+  revisited2process <-
+    atomicModifyIORef'
+      toRevisitSeqRef
+      (Seq.partition (\(WeightedOrderWithTimestamp _ _ oTime) -> diffUTCTime currentTime oTime < orderExecTime))
```

Comment:
Should be done passively as well as processing of suspended orders.

Reply:
toRevisit is seq. It wouldn't be efficient to use such approach